### PR TITLE
Update torchdata.nodes docs, use sphinx for API

### PR DESCRIPTION
**Pull request by andrewkho**
Use sphinx for torchdata.nodes api reference
split existing torchdata.nodes into "What is torchdata.nodes?" and "getting started"

This looks like a lot of line changes but it's mostly docstring updates, and splitting the existing docfile into two

Tested locally:

API page:
image

Getting started:
image

Migrating from torch.utils.data
image

What is torchdata.nodes?
image

### CODE REVIEW

**Review by divyanshk — APPROVED** [review]
lets goo!!!!

**Comment by divyanshk** [comment]
@andrewkho actually since you have this PR already open, do you want to add a quick section about perf in migrate_to_nodes_from_utils.rst ?

**Comment by divyanshk** [comment]
@andrewkho haven't you seen the news :-P all meta products are down

**Comment by divyanshk** [comment]
@andrewkho we discussed yesterday to include some perf related info in the docs, so that all the "informational pieces" are visible together in the docs. yeah we can include a pointer to the readme, or include a brief blurb here as well. up to you?

File: torchdata/nodes/base_node.py [hunk 1]
```diff
@@ -14,54 +14,58 @@
...
-
-    Subclasses of base node must implement the following methods:
-
-    * __init__() - must call super().__init__()
-    * reset(initial_state: Optional[dict]=None) - As above. Reset is a good place to put expensive
-        initialization, as it will be lazily called when next() or state_dict() is called.
-        Must call super().reset(initial_state)
-    * next() -> T - logic for returning the next value in the sequence, or throw StopIteration
-    * get_state(self) -> dict: returns a dictionary representing state that may be passed to reset()
+    """BaseNodes are the base class for creating composable dataloading dags in ``torchdata.nodes``.
```
Comment (ramanishsingh) [inline torchdata/nodes/base_node.py:17]: Question, should it be DAGs?

File: docs/source/migrate_to_nodes_from_utils.rst [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+==========================================================
+
+This guide is intended to help people familiar with ``torch.utils.data``, or
+:class:`~torchdata.stateful_dataloader.StatefulDataLoader`,
+to get started with ``torchdata.nodes``, and provide a starting ground for defining
+your own dataloading pipelines.
+
+We'll demonstrate how to achieve the most common DataLoader features, re-use existing samplers and datasets,
+and load/save dataloader state. It performs at least as well as ``DataLoader`` and ``StatefulDataLoader``,
+see :ref:`how-does-nodes-perform`.
```
Comment (andrewkho) [inline docs/source/migrate_to_nodes_from_utils.rst:13]: @divyanshk added xref here

File: docs/source/what_is_torchdata_nodes.rst [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Dataset sharding is required for data-parallel training, which is fairly
+reasonable. But what about sharding between dataloader workers? With
+Map-style datasets, distribution of work between workers is handled by
+the main process, which distributes sampler indices to workers. With
+IterableDatasets, each worker needs to figure out (through
+``torch.utils.data.get_worker_info``) what data it should be returning.
+
+.. _how-does-nodes-perform:
```
Comment (andrewkho) [inline docs/source/what_is_torchdata_nodes.rst:91]: @divyanshk it links here

File: docs/source/migrate_to_nodes_from_utils.rst [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            node = tn.PinMemory(node)
+        node = tn.Prefetcher(node, prefetch_factor=num_workers * 2)
+
+        # Note that node is an iterator, and once it's exhausted, you'll need to call .reset()
+        # on it to start a new Epoch.
+        # Insteaad, we wrap the node in a Loader, which is an iterable and handles reset. It
+        # also provides state_dict and load_state_dict methods.
+        return tn.Loader(node)
+
+Now let's test this out with a useless dataset, and demonstrate how state management works.
```
Comment (ramanishsingh) [inline docs/source/migrate_to_nodes_from_utils.rst:122]: nit: replace `useless` with `simple` or something else :D ?

File: torchdata/nodes/base_node.py (hunk 1 above)
Comment (andrewkho) [reply torchdata/nodes/base_node.py:17]: I think it's pretty un-ambiguous but you're correct, i'll update it

**Review by ramanishsingh — APPROVED** [review]

**Pull request merged**